### ISSUE TITLE
Saved model evaluate and validation always giving 0.50 accuracy despite reaching 97% on exact same data during training

### ISSUE THREAD

**Issue by aliencaocao** [keras-team/keras#15813]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): windows 10 21h2
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):  binary
- TensorFlow version (use command below):  v2.7.0-rc1-69-gc256c071bb2 2.7.0
- Python version: 3.9.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.3/8.2.1
- GPU model and memory: RTX 3080Ti 12GB

**Describe the current behavior**
When training a normal CNN model using the horse VS human dataset (binary image classification), the training accuracy reached 0.96+, but validation is forever stuck at 0.5. What's more interesting is after I load the saved model and called `model.evaluate` on the training set, it also gives 0.50, despite the training set being exactly same and was yielding 0.97+ during the actual training. This is actually a programming test project where I have to submit the model weight to google's server for their evaluation, and their evaluation also showed a >90% score.

**Describe the expected behavior**
model to perform normally in both training and testing

model weight: https://drive.google.com/file/d/11oqaGvy4vV0v77WmWzQxhtQSgEH4PUhA/view?usp=sharing

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```
import tensorflow as tf
import urllib
import zipfile
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.layers import *


def download_data():
    _TRAIN_URL = "https://storage.googleapis.com/download.tensorflow.org/data/horse-or-human.zip"
    _TEST_URL = "https://storage.googleapis.com/download.tensorflow.org/data/validation-horse-or-human.zip"
    urllib.request.urlretrieve(_TRAIN_URL, 'horse-or-human.zip')
    local_zip = 'horse-or-human.zip'
    zip_ref = zipfile.ZipFile(local_zip, 'r')
    zip_ref.extractall('tmp/horse-or-human/')
    zip_ref.close()
    urllib.request.urlretrieve(_TEST_URL, 'testdata.zip')
    local_zip = 'testdata.zip'
    zip_ref = zipfile.ZipFile(local_zip, 'r')
    zip_ref.extractall('tmp/testdata/')
    zip_ref.close()


def solution_model():
    train_datagen = ImageDataGenerator(rescale=1. / 255,
                                       rotation_range=40,
                                       width_shift_range=0.2,
                                       height_shift_range=0.2,
                                       shear_range=0.2,
                                       zoom_range=0.2,
                                       horizontal_flip=True,
                                       fill_mode='nearest')
    test_datagen = ImageDataGenerator(rescale=1. / 255)

    train_generator = train_datagen.flow_from_directory(
        'tmp/horse-or-human',  # This is the source directory for training images
        target_size=(300, 300),
        batch_size=16,
        class_mode='binary')

    print('Validation data')
    validation_generator = test_datagen.flow_from_directory(
        'tmp/testdata',
        target_size=(300, 300),
        batch_size=32)

    opt = tf.keras.optimizers.Adam(learning_rate=1e-3, epsilon=1e-6)
    loss = tf.keras.losses.BinaryCrossentropy(from_logits=True)
    epoch = 20
    batch_size = 16
    callbacks = [
        tf.keras.callbacks.EarlyStopping(monitor='val_accuracy', min_delta=0, patience=5, verbose=1,
                                         mode='auto', baseline=None, restore_best_weights=True),
        tf.keras.callbacks.ReduceLROnPlateau(monitor='val_accuracy', factor=0.05, patience=5, verbose=1)
    ]

    xInput = Input([300, 300, 3])
    x = tf.cast(xInput, tf.float32)
    resnet50 = tf.keras.applications.efficientnet.EfficientNetB0(include_top=False, weights='imagenet')
    x = resnet50(x)
    x = Flatten()(x)
    x = Dropout(0.2)(x)
    x = Dense(256)(x)
    x = BatchNormalization(epsilon=1.001e-5)(x)
    x = Activation('relu')(x)
    xOutput = Dense(1)(x)  # from logits so no need activation
    model = tf.keras.models.Model(xInput, xOutput)

    model.compile(optimizer=opt, loss=loss, metrics='accuracy')
    model.summary()
    model.fit(train_generator, validation_data=validation_generator, steps_per_epoch=train_generator.samples // batch_size, epochs=epoch, callbacks=callbacks, verbose=1)
    return model



if __name__ == '__main__':
    # download_data()  # only need to run this once
    model = solution_model()
    model.save("mymodel.h5")
```

Evaluation code:

```
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.models import load_model

test_datagen = ImageDataGenerator(rescale=1. / 255)

print('Validation data')
validation_generator = test_datagen.flow_from_directory(
    'tmp/horse-or-human',
    target_size=(300, 300),
    batch_size=32,
    class_mode='categorical')

model = load_model('mymodel.h5')
model.summary()
model.evaluate(validation_generator)  # should be >0.97
```

**Other info / logs** Include any logs or source code that would be helpful to
`33/33 [==============================] - 13s 125ms/step - loss: 0.7544 - accuracy: 0.5000`

**Comment by jvishnuvardhan** [comment]
@aliencaocao This is not a bug. The discrepancy is due to the `base_model` used in the model development. When i replaced `EfficientNetB0` with `MobileNetV2`, it gave very high accuracy for training and validation. Also, gave very high accuracy after reloading the model. 

The discrepancy could be due to layers (such as normalization layer) used in the model. Another thing is try to freeze the base model as the data is very small. Alternatively, you could also update architecture.

[Here](https://colab.research.google.com/gist/jvishnuvardhan/757d8abaaf3379031042770b23ddd7d9/untitled1151.ipynb) is a gist for reference. Thanks

Please verify once close the issue. [TF Forum](https://discuss.tensorflow.org/tags) or stackoverflow is better place to discuss more about this model architecture.

**Comment by aliencaocao** [comment]
I do not understand. If other models work with the same dataset but the efficient net will not, why is this behaviour normal and not a bug? Even if you said the normalization layers are different, it should not be affecting in a way that is reliably reproducible and produces exactly same behaviour of validation accuracy stuck at 50% sharp.

I have tried freezing the effecient net model, still same behaviour. val accuracy is forever 50% no matter train accuracy. Even when train accuracy is far worse at 39%, val accuracy is still 50%. How can this be a normal behaviour? Val loss is constantly decreasing normally, showing that the model state did change. I really think this is a bug where in the evaluation part of the model. Also, after I froze the effecient net model, train accuracy seem to never get past 50%. It will only get past 50% if I do not freeze the model. Thus I think freezing the model is not a proper solution.

![image](https://user-images.githubusercontent.com/20109683/152468345-9e30b61a-2c2c-4c8a-a083-502dbc6831bf.png)

I tried efficient net B1 also, both with layers frozen and unfrozen, both exactly same behaviour of val stuck at 50%.
![image](https://user-images.githubusercontent.com/20109683/152468831-d471569d-b58d-445a-ac45-191eebfc9894.png)


Moreover, I have used the exact same model on another dataset (oxford flowers 102) and it is working properly. This degree of uncertainty and data-picking behaviour is not something that I would expect from a mature model implementation. Other models working properly with this dataset does not mean the non-working one is free of bugs. This is not the first time I encountered this and I am not the only one also.

**Comment by jvishnuvardhan** [comment]
@aliencaocao `EfficientNetB0` architecture has builtin normalization layers, so we don't need to normalize the inputs as you did in the following part of the code. So, if you removed the normalization, then everything works as expected.

```
    train_datagen = ImageDataGenerator(rescale=1. / 255,
                                       rotation_range=40,
                                       width_shift_range=0.2,
                                       height_shift_range=0.2,
                                       shear_range=0.2,
                                       zoom_range=0.2,
                                       horizontal_flip=True,
                                       fill_mode='nearest')
    test_datagen = ImageDataGenerator(rescale=1. / 255)
```

Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/e226a9db905b57304b92b1187935e619/untitled1151.ipynb). Thanks!
